src/pykeen/triples/triples_factory.py

```diff
@@ -146,59 +146,465 @@ def _ensure_ids(
 
-    #: The inverse mapping for entity_label_to_id; initialized automatically
-    entity_id_to_label: Mapping[int, str] = dataclasses.field(init=False)
 
-    #: The inverse mapping for relation_label_to_id; initialized automatically
-    relation_id_to_label: Mapping[int, str] = dataclasses.field(init=False)
+@dataclasses.dataclass
+class CoreTriplesFactory:
+    """Create instances from ID-based triples."""
 
-    #: A vectorized version of entity_label_to_id; initialized automatically
-    _vectorized_entity_mapper: Callable[..., np.ndarray] = dataclasses.field(init=False)
+    def __init__(
+        self,
+        mapped_triples: MappedTriples,
+        num_entities: int,
+        num_relations: int,
+        create_inverse_triples: bool = False,
+        metadata: Optional[Mapping[str, Any]] = None,
+    ):
+        """
+        Create the triples factory.
+
+        :param mapped_triples: shape: (n, 3)
+            A three-column matrix where each row are the head identifier, relation identifier, then tail identifier.
+        :param num_entities:
+            The number of entities.
+        :param num_relations:
+            The number of relations.
+        :param create_inverse_triples:
+            Whether to create inverse triples.
+        :param metadata:
+            Arbitrary metadata to go with the graph
+        """
+        super().__init__()
+        self.mapped_triples = mapped_triples
+        self._num_entities = num_entities
+        self._num_relations = num_relations
+        self.create_inverse_triples = create_inverse_triples
+        if metadata is None:
+            metadata = dict()
+        self.metadata = metadata
+
+    @classmethod
+    def create(
+        cls,
+        mapped_triples: MappedTriples,
+        num_entities: Optional[int] = None,
+        num_relations: Optional[int] = None,
+        create_inverse_triples: bool = False,
+        metadata: Optional[Mapping[str, Any]] = None,
+    ) -> "CoreTriplesFactory":
+        """
+        Create a triples factory without any label information.
 
-    #: A vectorized version of relation_label_to_id; initialized automatically
-    _vectorized_relation_mapper: Callable[..., np.ndarray] = dataclasses.field(init=False)
+        :param mapped_triples: shape: (n, 3)
+            The ID-based triples.
+        :param num_entities:
+            The number of entities. If not given, inferred from mapped_triples.
+        :param num_relations:
+            The number of relations. If not given, inferred from mapped_triples.
+        :param create_inverse_triples:
+            Whether to create inverse triples.
+        :param metadata:
+            Additional metadata to store in the factory.
 
-    #: A vectorized version of entity_id_to_label; initialized automatically
-    _vectorized_entity_labeler: Callable[..., np.ndarray] = dataclasses.field(init=False)
+        :return:
+            A new triples factory.
+        """
+        if num_entities is None:
```

Comment:
why are these two none checks not directly in `__init__`?

Reply:
It enforces the user either to explicitly provide `num_entities` / `num_relations`, or use the `create` factory method.